presto-main/src/main/java/com/facebook/presto/execution/scheduler/nodeSelection/SimpleNodeSelector.java

```diff
@@ -111,19 +114,22 @@ public SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTas
         NodeMap nodeMap = this.nodeMap.get().get();
         NodeAssignmentStats assignmentStats = new NodeAssignmentStats(nodeTaskMap, nodeMap, existingTasks);
 
-        ResettableRandomizedIterator<InternalNode> randomCandidates = getRandomCandidates(maxTasksPerStage, nodeMap, existingTasks);
+        NodeSelection randomNodeSelection = new RandomNodeSelection(nodeMap, includeCoordinator, minCandidates, maxTasksPerStage, existingTasks);
         Set<InternalNode> blockedExactNodes = new HashSet<>();
         boolean splitWaitingForAnyNode = false;
+        List<HostAddress> sortedCandidates = NodeSelector.sortedNodes(nodeMap);
         for (Split split : splits) {
-            randomCandidates.reset();
-
             List<InternalNode> candidateNodes;
-            if (!split.isRemotelyAccessible()) {
-                candidateNodes = selectExactNodes(nodeMap, split.getAddresses(), includeCoordinator);
-            }
-            else {
-                candidateNodes = selectNodes(minCandidates, randomCandidates);
+            switch (split.getNodeSelectionStrategy()) {
+                case HARD_AFFINITY:
```

Comment:
I think in `HARD_AFFINITY` situation, `getPreferredNodes` is not a good name because `preferred` indicates it is not mandatory. 

In fact, `nodeSelectionStrategy` and `preferredNodes` are so coupled that I was thinking if we should return an object with a boolean and a list. The boolean would indicate whether the node list is a preference (best effort, for soft affinity) or requirement (for hard affinity). Then I think we probably don't even need to expose `getNodeSelectionStrategy`. WDYT? cc @highker

Reply:
I agree with a single interface can make things easier to understand. We can have two interfaces for this PR I guess and move to a unified interface in the next one.